Add support for unbounded repetition terms in the new property parser generator

weinig opened the pull request:
#### 19ef95a624848a4b52b40c5c664357a3993fccc6
<pre>
Add support for unbounded repetition terms in the new property parser generator
<a href="https://bugs.webkit.org/show_bug.cgi?id=248355">https://bugs.webkit.org/show_bug.cgi?id=248355</a>
rdar://102675501

Reviewed by Darin Adler.

Adds support for parsing unbounded comma separated lists via CSSProperties.json.
Specifically, this adds support for the `#` operator without range refinement
(e.g. no #{1,4}):

  &quot;&quot;&quot;
   A hash mark (#) indicates that the preceding type, word, or group occurs
   one or more times, separated by comma tokens (which may optionally be
   surrounded by white space and/or comments).
  &quot;&quot;&quot;
    -  <a href="https://www.w3.org/TR/css-values-4/#mult-comma">https://www.w3.org/TR/css-values-4/#mult-comma</a>

This brings with it support for a large collection (29) of longhands in the
animation, transition, background and mask families of properties.

To generate the matching for one of these lists, a new term type can be
specified in CSSProperties.json with &apos;kind&apos; set to &apos;repetition&apos; and &apos;value&apos;
set to whatever should be repeatedly matched. As a shorthand, the spec syntax
utilizing a `#` can also be used. For instance, &apos;animation-duration&apos;, which
has a specified grammar of:

  `&lt;&apos;animation-duration&apos;&gt; = &lt;time [0s,∞]#`

uses the following:

  &quot;parser-grammar&quot;: &quot;&lt;time [0,inf]&gt;#&quot;

* Source/WebCore/css/CSSProperties.json:
Replaces use of custom-parser/parser-function with parser-grammar for the
newly supported properties and adds a set of new shared grammar rules for
implementing them.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeCounterStyleRange): Adopt new consumeCommaSeparatedList* helper.
(WebCore::consumeCounterStyleAdditiveSymbols): Ditto.
(WebCore::consumeOverrideColorsDescriptor): Ditto.
(WebCore::isValidAnimationPropertyList): Moved from CSSPropertyParserHelpers as
now the only caller is in this file.
(WebCore::consumeAnimationValueForShorthand): Ditto. Also now implemented using
more exported shared grammar rules.
(WebCore::CSSPropertyParser::consumeAnimationShorthand): Adopt new
consumeCommaSeparatedList* helper
(WebCore::addBackgroundValue): Moved from CSSPropertyParserHelpers as now the only
caller is in this file.
(WebCore::CSSPropertyParser::consumeBackgroundShorthand): Split background-size and
mask size as they now have separate consume functions (though at the moment identical).

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFontVariationSettings): Adopt new
consumeCommaSeparatedList* helper.
(WebCore::CSSPropertyParserHelpers::consumeFontFamily): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeKeyframesName): Added to allow &quot;animation-name&quot;
to be mostly generated. Unused CSSParserContext parameteter is a by product consumers called
from generated consumers are expected to take a CSSParserContext for the moment. This should
be refined to not be required in the future.
(WebCore::CSSPropertyParserHelpers::consumeSingleTransitionPropertyIdent):
(WebCore::CSSPropertyParserHelpers::consumeSingleTransitionPropertyOrNone):
(WebCore::CSSPropertyParserHelpers::consumeSingleTransitionProperty):
Split out shorthand and longhand behaviors into separate functions (with a shared
helper between them). Previously, the shorthand version (allowing &quot;none&quot;) was used
in both places, and then a post consume check if any values were &quot;none&quot; was done.
Now that we are generating the repetition, doing the post consume check is non-trivial
so just doing the parse right makes more sense.
(WebCore::CSSPropertyParserHelpers::consumeSteps): Make static (and remove forward
declaration from header. Nothing outside this file uses this function.
(WebCore::CSSPropertyParserHelpers::consumeCubicBezier): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSpringFunction): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeTimingFunction): Renamed from
consumeAnimationTimingFunction to better reflect the spec naming and to indicate it used
for more than just animation as it transtions also use it.
(WebCore::CSSPropertyParserHelpers::consumeShadow): Make static (and remove forward
declaration from header. Nothing outside this file uses this function.
(WebCore::CSSPropertyParserHelpers::consumeScrollSnapType): Use auto.
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeOrBox): Use a generated helper and
add a note about the spec ambiguity along with link to spec github issue filed:
<a href="https://github.com/w3c/fxtf-drafts/issues/481">https://github.com/w3c/fxtf-drafts/issues/481</a>
(WebCore::CSSPropertyParserHelpers::consumeBaselineKeyword): Make static (and remove forward
declaration from header. Nothing outside this file uses this function
(WebCore::CSSPropertyParserHelpers::consumeBackgroundSize): Make this a template function
generic on CSSPropertyID as the callers all statically know which variant they want.
(WebCore::CSSPropertyParserHelpers::consumeRepeatStyle): Make static (and remove forward
declaration from header. Nothing outside this file uses this function
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundRepeat): Added to make it a bit
more clear in consumeBackgroundComponent() what the behavior for each property is.
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundPositionX): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundPositionY): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundSize): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleMaskRepeat): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleMaskSize): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleWebkitBackgroundSize): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleWebkitMaskPositionX): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleWebkitMaskPositionY): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeBackgroundComponent): Reordered to group families
of properties together and updated to use generated consumers where possible or the new
named consumers above.
(WebCore::CSSPropertyParserHelpers::consumeCommaSeparatedBackgroundComponent): Adopt new
consumeCommaSeparatedList* helper

(WebCore::CSSPropertyParserHelpers::consumeAutoOrLengthOrPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeMarginSide): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSide): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAnimationIterationCount): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAnimationName): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeTransitionProperty): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAnimationTimingFunction): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAnimationValue): Deleted.
(WebCore::CSSPropertyParserHelpers::isValidAnimationPropertyList): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAnimationPropertyList): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeBackgroundBlendMode): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeBackgroundAttachment): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeBackgroundBox): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeMaskClip): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeBackgroundClip): Deleted.
(WebCore::CSSPropertyParserHelpers::consumePrefixedMaskComposite): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeMaskComposite): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeWebkitMaskSourceType): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeWebkitMaskMode): Deleted.
(WebCore::CSSPropertyParserHelpers::consumePrefixedBackgroundBox): Deleted.
(WebCore::CSSPropertyParserHelpers::addBackgroundValue): Deleted.
Replaced by generated shared grammar rules or generated properties..

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
(WebCore::CSSPropertyParserHelpers::consumeCommaSeparatedListWithSingleValueOptimization):
(WebCore::CSSPropertyParserHelpers::consumeCommaSeparatedListWithoutSingleValueOptimization):
Add two helpers which can consume a comma separated list, one with an optimization to return
the underlying CSSValue if there was only one element in the list (and a CSSValueList otherwise)
and the other which always returns a CSSValueList. This allows callers to maintain their
existing behavior.

* Source/WebCore/css/process-css-properties.py:
(Term.from_json): Create a RepetitionTerm if the &apos;kind&apos; is &apos;repetition&apos;.
(RepetitionTerm): RepetitionTerm has a unique &apos;single-value-optimization&apos; flag
that can be set on it to indicate that it should the variant of consumeCommaSeparatedList*
that does the single value optimizaton.
(TermGeneratorRepetitionTerm:) To generate code for a RepetitionTerm, we utilize one of the
consumeCommaSeparatedList* functions and pass it a lambda which contains the submatch. The
new automatic indentation support really came in handy here to allow an arbitrarty set of
terms to generate inside the lambda without issue.

Canonical link: <a href="https://commits.webkit.org/257051@main">https://commits.webkit.org/257051@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc6dfe00035f0997efda524e6a7ca8839b0eb704

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97722 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6966 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30890 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107214 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167481 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101668 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7375 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35734 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90109 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103865 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103365 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5535 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84354 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/32502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/101280 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87409 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89164 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/958 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/948 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22094 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4846 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5769 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44556 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2203 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41486 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -1101,9 +1101,10 @@
-                "parser-requires-context": true
+                "parser-grammar": {
+                    "value": "<single-animation-composition>#",
+                    "single-value-optimization": true
```
Let’s reverse the sense of this and have it default to true?

I feel that soon we’ll be changing what kinds of objects we parse into. It’s not important in the long run for us to have many different ways to represent the same parsed thing. List containing one item vs. item not in a list is one. Value list, pair, quad is another. Also the thing where we parse and let the parser implement the rules for filling in missing values, vs. doing that in the style builder. It’s not clear to me to what extent any of these are web-exposed. We will probably want to optimized for memory use and perhaps create idioms so the style builder and computed style code can be written elegantly and economically.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -1373,6 +1419,21 @@ bool CSSPropertyParser::consumeAnimationShorthand(const StylePropertyShorthand&
+{
+    if (list) {
+        if (!list->isBaseValueList()) {
+            RefPtr<CSSValue> firstValue = list;
```
```
auto firstValue = list.releaseNonNull();
```

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -1327,6 +1326,53 @@ bool CSSPropertyParser::consumeLegacyTextOrientation(bool important)
+    if (property != CSSPropertyTransitionProperty || valueList.length() < 2)
+        return true;
+    for (auto& value : valueList) {
+        if (value->isPrimitiveValue() && downcast<CSSPrimitiveValue>(value.get()).isValueID()
```
The `isValueID` check here is unneeded. Further, I think there is a helper that does this operation on a CSSValue, checking if it’s a particular keyword, but I think maybe in another source file.

weinig replied:
Found the one I think you mean in StyleProperties.cpp and moved it to a shared location.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -5427,18 +5420,13 @@ static RefPtr<CSSValue> consumeGenericFamily(CSSParserTokenRange& range)
+            return parsedValue;
+        if (auto parsedValue = consumeFamilyName(range))
+            return parsedValue;
+        return nullptr;
```
Or just return `consumeFamilyName`.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -5684,22 +5666,50 @@ RefPtr<CSSValue> consumeAnimationName(CSSParserTokenRange& range)
-RefPtr<CSSValue> consumeTransitionProperty(CSSParserTokenRange& range)
+static RefPtr<CSSValue> consumeSingleTransitionPropertyIdent(CSSParserTokenRange& range, const CSSParserToken& token)
 {
+    if (CSSPropertyID property = token.parseAsCSSPropertyID()) {
```
auto

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -5684,22 +5666,50 @@ RefPtr<CSSValue> consumeAnimationName(CSSParserTokenRange& range)
+    // "The <custom-ident> production in <single-transition-property> also excludes the keyword
+    //  none, in addition to the keywords always excluded from <custom-ident>."
+
+    const CSSParserToken& token = range.peek();
```
`auto&`

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
```diff
@@ -413,6 +396,42 @@ Ref<CSSPrimitiveValue> createPrimitiveValuePair(Args&&... args)
+{
+    RefPtr<CSSValue> result;
+    do {
+        RefPtr<CSSValue> value = std::invoke(subConsumer, range, std::forward<Args>(args)...);
```
Can this be just `RefPtr` or `auto`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
```diff
@@ -413,6 +396,42 @@ Ref<CSSPrimitiveValue> createPrimitiveValuePair(Args&&... args)
+            return nullptr;
+        if (result) {
+            if (!result->isBaseValueList()) {
+                RefPtr<CSSValue> firstValue = WTFMove(result);
```
```
auto firstValue = list.releaseNonNull();
```

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
```diff
@@ -413,6 +396,42 @@ Ref<CSSPrimitiveValue> createPrimitiveValuePair(Args&&... args)
+{
+    auto result = CSSValueList::createCommaSeparated();
+    do {
+        RefPtr<CSSValue> value = std::invoke(subConsumer, range, std::forward<Args>(args)...);
```
Can this be just `RefPtr` or `auto`?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -1373,6 +1419,21 @@ bool CSSPropertyParser::consumeAnimationShorthand(const StylePropertyShorthand&
     return m_range.atEnd();
 }
 
+static void addBackgroundValue(RefPtr<CSSValue>& list, Ref<CSSValue>&& value)
```
Any way to share the code with `consumeCommaSeparatedListWithSingleValueOptimization`?

Commit:
Add support for unbounded repetition terms in the new property parser generator
https://bugs.webkit.org/show_bug.cgi?id=248355
rdar://102675501

Reviewed by Darin Adler.

Adds support for parsing unbounded comma separated lists via CSSProperties.json.
Specifically, this adds support for the `#` operator without range refinement
(e.g. no #{1,4}):

  """
   A hash mark (#) indicates that the preceding type, word, or group occurs
   one or more times, separated by comma tokens (which may optionally be
   surrounded by white space and/or comments).
  """
    -  https://www.w3.org/TR/css-values-4/#mult-comma

This brings with it support for a large collection (29) of longhands in the
animation, transition, background and mask families of properties.

To generate the matching for one of these lists, a new term type can be
specified in CSSProperties.json with 'kind' set to 'repetition' and 'value'
set to whatever should be repeatedly matched. As a shorthand, the spec syntax
utilizing a `#` can also be used. For instance, 'animation-duration', which
has a specified grammar of:

  `<'animation-duration'> = <time [0s,∞]#`

uses the following:

  "parser-grammar": "<time [0,inf]>#"

* Source/WebCore/css/CSSProperties.json:
Replaces use of custom-parser/parser-function with parser-grammar for the
newly supported properties and adds a set of new shared grammar rules for
implementing them.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeCounterStyleRange): Adopt new consumeCommaSeparatedList* helper.
(WebCore::consumeCounterStyleAdditiveSymbols): Ditto.
(WebCore::consumeOverrideColorsDescriptor): Ditto.
(WebCore::isValidAnimationPropertyList): Moved from CSSPropertyParserHelpers as
now the only caller is in this file.
(WebCore::consumeAnimationValueForShorthand): Ditto. Also now implemented using
more exported shared grammar rules.
(WebCore::CSSPropertyParser::consumeAnimationShorthand): Adopt new
consumeCommaSeparatedList* helper
(WebCore::addBackgroundValue): Moved from CSSPropertyParserHelpers as now the only
caller is in this file.
(WebCore::CSSPropertyParser::consumeBackgroundShorthand): Split background-size and
mask size as they now have separate consume functions (though at the moment identical).

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFontVariationSettings): Adopt new
consumeCommaSeparatedList* helper.
(WebCore::CSSPropertyParserHelpers::consumeFontFamily): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeKeyframesName): Added to allow "animation-name"
to be mostly generated. Unused CSSParserContext parameteter is a by product consumers called
from generated consumers are expected to take a CSSParserContext for the moment. This should
be refined to not be required in the future.
(WebCore::CSSPropertyParserHelpers::consumeSingleTransitionPropertyIdent):
(WebCore::CSSPropertyParserHelpers::consumeSingleTransitionPropertyOrNone):
(WebCore::CSSPropertyParserHelpers::consumeSingleTransitionProperty):
Split out shorthand and longhand behaviors into separate functions (with a shared
helper between them). Previously, the shorthand version (allowing "none") was used
in both places, and then a post consume check if any values were "none" was done.
Now that we are generating the repetition, doing the post consume check is non-trivial
so just doing the parse right makes more sense.
(WebCore::CSSPropertyParserHelpers::consumeSteps): Make static (and remove forward
declaration from header. Nothing outside this file uses this function.
(WebCore::CSSPropertyParserHelpers::consumeCubicBezier): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSpringFunction): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeTimingFunction): Renamed from
consumeAnimationTimingFunction to better reflect the spec naming and to indicate it used
for more than just animation as it transtions also use it.
(WebCore::CSSPropertyParserHelpers::consumeShadow): Make static (and remove forward
declaration from header. Nothing outside this file uses this function.
(WebCore::CSSPropertyParserHelpers::consumeScrollSnapType): Use auto.
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeOrBox): Use a generated helper and
add a note about the spec ambiguity along with link to spec github issue filed:
https://github.com/w3c/fxtf-drafts/issues/481
(WebCore::CSSPropertyParserHelpers::consumeBaselineKeyword): Make static (and remove forward
declaration from header. Nothing outside this file uses this function
(WebCore::CSSPropertyParserHelpers::consumeBackgroundSize): Make this a template function
generic on CSSPropertyID as the callers all statically know which variant they want.
(WebCore::CSSPropertyParserHelpers::consumeRepeatStyle): Make static (and remove forward
declaration from header. Nothing outside this file uses this function
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundRepeat): Added to make it a bit
more clear in consumeBackgroundComponent() what the behavior for each property is.
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundPositionX): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundPositionY): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleBackgroundSize): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleMaskRepeat): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleMaskSize): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleWebkitBackgroundSize): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleWebkitMaskPositionX): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeSingleWebkitMaskPositionY): Ditto.
(WebCore::CSSPropertyParserHelpers::consumeBackgroundComponent): Reordered to group families
of properties together and updated to use generated consumers where possible or the new
named consumers above.
(WebCore::CSSPropertyParserHelpers::consumeCommaSeparatedBackgroundComponent): Adopt new
consumeCommaSeparatedList* helper

(WebCore::CSSPropertyParserHelpers::consumeAutoOrLengthOrPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeMarginSide): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeSide): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAnimationIterationCount): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAnimationName): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeTransitionProperty): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAnimationTimingFunction): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAnimationValue): Deleted.
(WebCore::CSSPropertyParserHelpers::isValidAnimationPropertyList): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeAnimationPropertyList): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeBackgroundBlendMode): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeBackgroundAttachment): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeBackgroundBox): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeMaskClip): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeBackgroundClip): Deleted.
(WebCore::CSSPropertyParserHelpers::consumePrefixedMaskComposite): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeMaskComposite): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeWebkitMaskSourceType): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeWebkitMaskMode): Deleted.
(WebCore::CSSPropertyParserHelpers::consumePrefixedBackgroundBox): Deleted.
(WebCore::CSSPropertyParserHelpers::addBackgroundValue): Deleted.
Replaced by generated shared grammar rules or generated properties..

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
(WebCore::CSSPropertyParserHelpers::consumeCommaSeparatedListWithSingleValueOptimization):
(WebCore::CSSPropertyParserHelpers::consumeCommaSeparatedListWithoutSingleValueOptimization):
Add two helpers which can consume a comma separated list, one with an optimization to return
the underlying CSSValue if there was only one element in the list (and a CSSValueList otherwise)
and the other which always returns a CSSValueList. This allows callers to maintain their
existing behavior.

* Source/WebCore/css/process-css-properties.py:
(Term.from_json): Create a RepetitionTerm if the 'kind' is 'repetition'.
(RepetitionTerm): RepetitionTerm has a unique 'single-value-optimization' flag
that can be set on it to indicate that it should the variant of consumeCommaSeparatedList*
that does the single value optimizaton.
(TermGeneratorRepetitionTerm:) To generate code for a RepetitionTerm, we utilize one of the
consumeCommaSeparatedList* functions and pass it a lambda which contains the submatch. The
new automatic indentation support really came in handy here to allow an arbitrarty set of
terms to generate inside the lambda without issue.

Canonical link: https://commits.webkit.org/257051@main
